srml/im-online/src/tests.rs

```diff
@@ -0,0 +1,187 @@
+
+#![cfg(test)]
+
+use super::*;
+use crate::mock::*;
+use offchain::testing::TestOffchainExt;
+use primitives::offchain::OpaquePeerId;
+use runtime_io::with_externalities;
+use sr_primitives::testing::UintAuthorityId;
+
+
+#[test]
+fn test_unresponsiveness_slash_fraction() {
+	// A single case of unresponsiveness is not slashed.
+	assert_eq!(
+		UnresponsivenessOffence::<()>::slash_fraction(1, 50),
+		Perbill::zero(),
+	);
+
+	assert_eq!(
+		UnresponsivenessOffence::<()>::slash_fraction(3, 50),
+		Perbill::from_parts(6000000), // 0.6%
+	);
+
+	// One third offline should be punished around 5%.
+	assert_eq!(
+		UnresponsivenessOffence::<()>::slash_fraction(17, 50),
+		Perbill::from_parts(48000000), // 4.8%
+	);
+}
+
+#[test]
+fn should_report_offline_validators() {
+	with_externalities(&mut new_test_ext(), || {
+		// given
+		let block = 1;
+		System::set_block_number(block);
+		// buffer new validators
+		Session::rotate_session();
+		// enact the change and buffer another one
+		let validators = vec![1, 2, 3, 4, 5, 6];
+		VALIDATORS.with(|l| *l.borrow_mut() = Some(validators.clone()));
+		Session::rotate_session();
+
+		// when
+		// we end current session and start the next one
+		Session::rotate_session();
+
+		// then
+		let offences = OFFENCES.with(|l| l.replace(vec![]));
+		assert_eq!(offences, vec![
+			(vec![], UnresponsivenessOffence {
+				session_index: 2,
+				validator_set_count: 3,
+				offenders: vec![
+					(1, 1),
+					(2, 2),
+					(3, 3),
+				],
+			})
+		]);
+
+		// should not report when heartbeat is sent
+		for (idx, v) in validators.into_iter().take(4).enumerate() {
+			heartbeat(block, 3, idx as u32, v.into());
+		}
+		Session::rotate_session();
+
+		// then
+		let offences = OFFENCES.with(|l| l.replace(vec![]));
+		assert_eq!(offences, vec![
+			(vec![], UnresponsivenessOffence {
+				session_index: 3,
+				validator_set_count: 6,
+				offenders: vec![
+					(5, 5),
+					(6, 6),
+				],
+			})
+		]);
+	});
+}
+
+fn heartbeat(
+	block_number: u64,
+	session_index: u32,
+	authority_index: u32,
+	id: UintAuthorityId,
+) {
+	let heartbeat = Heartbeat {
+		block_number,
+		network_state: OpaqueNetworkState {
+			peer_id: OpaquePeerId(vec![1]),
+			external_addresses: vec![],
+		},
+		session_index,
+		authority_index,
+	};
+	let signature = id.sign(&heartbeat.encode()).unwrap();
+
+	// when
+	ImOnline::heartbeat(
+		Origin::system(system::RawOrigin::None),
+		heartbeat,
+		signature
+	).unwrap();
+}
+
+#[test]
+fn should_correctly_mark_online_validator_when_heartbeat_is_received() {
+	with_externalities(&mut new_test_ext(), || {
+		// given
+		let block = 1;
+		System::set_block_number(block);
+		// buffer new validators
+		Session::rotate_session();
+		// enact the change and buffer another one
+		VALIDATORS.with(|l| *l.borrow_mut() = Some(vec![1, 2, 3, 4, 5, 6]));
+		Session::rotate_session();
+		assert!(!ImOnline::is_online_in_current_session(0));
+
+		// when
+		heartbeat(block, 2, 0, 1.into());
```

Comment:
I don't get how in block 1 session_index = 2

Reply:
(I changed the code a bit but my question still holds)
(I am aware of the buffering and that we need one extra rotate_session to write the buffered validators)

Reply:
Session rotations are independent of blocks. In regular `node-runtime` that would never happen though.